File: test/python_impl_tests/test_struct/test_struct.py
#!/usr/bin/python

# Copyright 2021-2024 NXP
#
# SPDX-License-Identifier: BSD-3-Clause

import erpc
import pytest

from .service.erpc_outputs.test_ArithmeticService import common
from .service.erpc_outputs.test_ArithmeticService.client import ArithmeticService1Client
from .service.erpc_outputs.test_ArithmeticService.client import ArithmeticService2Client
from .service.erpc_outputs.test_ArithmeticService.server import ArithmeticService1Service
from .service.erpc_outputs.test_ArithmeticService.server import ArithmeticService2Service
from .service.erpc_outputs.test_ArithmeticService.interface import IArithmeticService1
from .service.erpc_outputs.test_ArithmeticService.interface import IArithmeticService2


###############################################################################
# Variables
###############################################################################


###############################################################################
# Fixtures
###############################################################################
@pytest.fixture()
def client1(client_manager):
    if client_manager is None:
        pytest.skip()

    return ArithmeticService1Client(client_manager)


@pytest.fixture()
def client2(client_manager):
    if client_manager is None:
        pytest.skip()

    return ArithmeticService2Client(client_manager)


@pytest.fixture(scope="module")
def services():
    return [
        ArithmeticService1Service(StructTestServiceHandler1()),
        ArithmeticService2Service(StructTestServiceHandler2())
    ]


###############################################################################
# Handlers
###############################################################################
class StructTestServiceHandler1(IArithmeticService1):
    def getMember(self, c):
        return c.m

    def sendNestedStruct(self, d):
        return d.c.m

    def returnStruct(self, x, y):
        b = common.B()
        b.x = x
        b.y = y
        return b

    def getMemberTest2(self, a):
        result = common.B()
        result.x = a.c.m
        result.y = a.b.x
        return result

    def checkString(self, p):
        return not (p.species == "Ape")

    def returnStrings(self):
        str_str = common.stringStruct(one="One", two="Two", three="Three")
        return str_str


class StructTestServiceHandler2(IArithmeticService2):
    def sendManyInts(self, f):
        if (f.a == -1) and (f.b == -1) and (f.c == -1) and (f.d == -1):
            return 0
        return 1

    def sendManyUInts(self, g):
        if (g.a == 0xff) and (g.b == 0xffff) and (g.c == 0xffffffff) and (g.d == 0xffffffffffffffff):
            return 0
        return 1

    def getStudentName(self, stud):
        return stud.name

    def getStudentTestAverage(self, stud):
        total = float(0.0)
        for x in range(3):
            total += stud.test_grades[x]
        return float(total / 3)

    def getStudentYear(self, stud):
        return stud.school_year

    def getStudentAge(self, stud):
        return stud.age

    def createStudent(self, name, test_scores, year, age):
        newStudent = common.student(name=name, test_grades=test_scores, school_year=year, age=age)
        return newStudent

    def test_struct_allDirection(self, a, b, e):
        e.value.number = a.number * b.number
        e.value.text = "World"
        e.value.color = common.enumColor.green
        e.value.c.m = a.c.m * b.c.m
        e.value.c.n = a.c.n * b.c.n
        for i in range(len(a.list_numbers)):
            e.value.list_numbers[i] = a.list_numbers[i] * b.list_numbers[i]
            e.value.list_text[i] = "World"
        for i in range(len(a.array_numbers)):
            e.value.array_numbers[i] = a.array_numbers[i] * b.array_numbers[i]
            e.value.array_text[i] = "World"
        for i in range(len(a.binary_numbers)):
            e.value.binary_numbers[i] = a.binary_numbers[i] * b.binary_numbers[i]
        return

    def testSendingByrefMembers(self, s):
        if (s.a == 0xED) and (s.b == 4):
            return True
        return False


###############################################################################
# Tests
###############################################################################
def test_GetMember1(client1):
    c = common.C(m=4, n=5)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    r = client1.getMember(c)
    assert r == 4


def test_TestArithmeticServiceID():
    assert 3 == IArithmeticService1.SERVICE_ID


def test_NestedStruct1(client1):
    c = common.C(m=4, n=5)
    d = common.D(c)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    r = client1.sendNestedStruct(d)
    assert r == 4


def test_ReturnStruct1(client1):
    b = common.B()
    # send request to the server
    print('\r\neRPC request is sent to the server')
    b = client1.returnStruct(float(3.14), float(2.71828))
    print(b)
    assert round(b.x, 5) == float(3.14)
    assert round(b.y, 5) == float(2.71828)


def test_SendIntSizes(client2):
    f = common.F(-1, -1, -1, -1)
    # print(f)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    assert 0 == client2.sendManyInts(f)


def test_SendUnsignedIntSizes(client2):
    g = common.G(0xFF, 0xFFFF, 0xFFFFFFFF, 0xFFFFFFFFFFFFFFFF)
    # print(g)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    assert 0 == client2.sendManyUInts(g)


def test_GetMember2(client1):
    c = common.C(m=4, n=5)
    b = common.B(x=float(3.14), y=float(2.71828))
    a = common.A(b=b, c=c)
    bb = common.B()
    # send request to the server
    print('\r\neRPC request is sent to the server')
    bb = client1.getMemberTest2(a)
    # print(bb)
    assert round(float(bb.x), 2) == float(4.0)
    assert round(float(bb.y), 2) == float(3.14)


def test_TestString1(client1):
    prim8 = common.primate(species="Ape", is_ape=False)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    assert 0 == client1.checkString(prim8)


def test_TestReturnString1(client1):
    result = common.stringStruct()
    # send request to the server
    print('\r\neRPC request is sent to the server')
    result = client1.returnStrings()
    assert result.one == "One"
    assert result.two == "Two"
    assert result.three == "Three"


def test_TestStudent1(client2):
    stud = common.student(name="Donnie Darko",
                          test_grades=[float(85.75), float(93.29), float(92.46)],
                          school_year=common.school_year_t.SENIOR, age=19)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    assert "Donnie Darko" == client2.getStudentName(stud)
    assert float(
        (stud.test_grades[0] + stud.test_grades[1] + stud.test_grades[2]) / 3) == client2.getStudentTestAverage(
        stud)
    assert 19 == client2.getStudentAge(stud)
    assert 12 == common.school_year_t.SENIOR
    assert common.school_year_t.SENIOR == client2.getStudentYear(stud)


def test_TestStudent2(client2):
    stud = common.student()
    # send request to the server
    print('\r\neRPC request is sent to the server')
    stud = client2.createStudent("George Bush", [float(65.32), float(69.68), float(0)],
                                 common.school_year_t.FRESHMAN, 68)
    assert "George Bush" == client2.getStudentName(stud)
    assert float((65.32 + 69.68 + 0) / 3) == client2.getStudentTestAverage(stud)
    assert 68 == client2.getStudentAge(stud)
    assert common.school_year_t.FRESHMAN == client2.getStudentYear(stud)


def test_struct_allDirection(client2):
    a_binary = bytearray(b'')
    b_binary = bytearray(b'')
    e_binary = bytearray(b'')
    for x in range(2):
        a_binary.append(x)
        b_binary.append(x * 2)
        e_binary.append(x)
    a = common.AllTypes(number=5, text="Hello",
                        color=common.enumColor.red,
                        c=common.C(m=5, n=20), list_numbers=[0, 1],
                        list_text=["Hello", "Hello"], array_numbers=[0, 1],
                        array_text=["Hello", "Hello"], binary_numbers=a_binary)
    b = common.AllTypes(number=10, text="World",
                        color=common.enumColor.green,
                        c=common.C(m=7, n=17), list_numbers=[1, 2],
                        list_text=["World", "World"], array_numbers=[1, 2],
                        array_text=["World", "World"], binary_numbers=b_binary)
    e = erpc.Reference(common.AllTypes(number=15, text="Hello",
                                       color=common.enumColor.blue,
                                       c=common.C(m=4, n=14),
                                       list_numbers=[2, 3], list_text=["Hello", "Hello"],
                                       array_numbers=[2, 3], array_text=["Hello", "Hello"],
                                       binary_numbers=e_binary))
    # send request to the server
    print('\r\neRPC request is sent to the server')
    client2.test_struct_allDirection(a, b, e)
    assert e.value.number == a.number * b.number
    assert e.value.text == "World"
    assert e.value.color == common.enumColor.green
    assert e.value.c.m == a.c.m * b.c.m
    assert e.value.c.n == a.c.n * b.c.n
    assert len(e.value.list_numbers) == 2
    for x in range(len(a.list_numbers)):
        assert e.value.list_numbers[x] == a.list_numbers[x] * b.list_numbers[x]
        assert "World" == e.value.list_text[x]
    for x in range(2):
        assert e.value.array_numbers[x] == a.array_numbers[x] * b.array_numbers[x]
        assert "World" == e.value.array_text[x]
    for x in range(len(a.binary_numbers)):
        assert e.value.binary_numbers[x] == a.binary_numbers[x] * b.binary_numbers[x]


@pytest.mark.skip("Not implemented")
def test_TestSendingByrefMembers(client2):
    s = common.StructWithByrefMembers()
    b = 4
    s.a = 0xED
    s.b = id(b)
    # send request to the server
    print('\r\neRPC request is sent to the server')
    assert True == client2.testSendingByrefMembers(s)


def test_Quit(common_client):
    # send request to the server
    print("eRPC quit request send to the server.")
    return common_client.quit()

File: test/java_impl_tests/src/test/java/io/github/embeddedrpc/erpc/tests/client/TestStruct.java
/*
 * Copyright 2023 NXP
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

package io.github.embeddedrpc.erpc.tests.client;


import io.github.embeddedrpc.erpc.auxiliary.Reference;
import io.github.embeddedrpc.erpc.tests.TestingClient;
import io.github.embeddedrpc.erpc.tests.test_arrays.test_unit_test_common.client.CommonClient;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.client.ArithmeticService1Client;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.client.ArithmeticService2Client;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.common.enums.enumColor;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.common.enums.school_year_t;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.common.structs.*;
import io.github.embeddedrpc.erpc.tests.test_struct.erpc_outputs.test_ArithmeticService.interfaces.IArithmeticService1;
import org.junit.jupiter.api.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class TestStruct extends TestingClient {
    private final ArithmeticService1Client client1 = new ArithmeticService1Client(clientManager);
    private final ArithmeticService2Client client2 = new ArithmeticService2Client(clientManager);

    @Test
    public void testArithmeticServiceID() {
        assertEquals(3, IArithmeticService1.SERVICE_ID);
    }

    @Test
    public void getMember1() {
        C c = new C(4, 5);

        int r = client1.getMember(c);

        assertEquals(4, r);
    }

    @Test
    public void nestedStruct1() {
        C c = new C(4, 5);
        D d = new D(c);

        int r = client1.sendNestedStruct(d);

        assertEquals(4, r);
    }

    @Test
    public void returnStruct1() {
        B b;

        b = client1.returnStruct(3.14F, 2.71828F);

        assertEquals(3.14F, b.get_x());
        assertEquals(2.71828F, b.get_y());
    }

    @Test
    public void sendIntSizes() {
        F f = new F((byte) -1, (short) -1, -1, (long) -1);

        assertEquals(0, client2.sendManyInts(f));
    }

    @Test
    public void sendUnsignedIntSizes() {
        // UInt64 not supported
    }

    @Test
    public void getMember2() {
        C c = new C(4, 5);
        B b = new B(3.14F, 2.71828F);
        A a = new A(b, c);
        B bb;

        bb = client1.getMemberTest2(a);

        assertEquals(4.0F, bb.get_x());
        assertEquals(3.14F, bb.get_y());
    }

    @Test
    public void testString1() {
        primate prim8 = new primate("Ape", false);

        assertEquals(0, client1.checkString(prim8));
    }

    @Test
    public void testReturnString1() {
        stringStruct result;

        result = client1.returnStrings();

        assertEquals("One", result.get_one());
        assertEquals("Two", result.get_two());
        assertEquals("Three", result.get_three());
    }

    @Test
    public void testStudent1() {
        student stud = new student("Donnie Darko", new float[]{85.75F, 93.29F, 92.46F}, school_year_t.SENIOR, 19);
        assertEquals("Donnie Darko", client2.getStudentName(stud));
        assertEquals((stud.get_test_grades()[0] + stud.get_test_grades()[1] + stud.get_test_grades()[2]) / 3, client2.getStudentTestAverage(stud));
        assertEquals(19, client2.getStudentAge(stud));
        assertEquals(school_year_t.SENIOR.getValue(), client2.getStudentYear(stud));
    }

    @Test
    public void testStudent2() {
        student stud;

        stud = client2.createStudent("George Bush", new float[]{65.32F, 69.68F, 0}, school_year_t.FRESHMAN, 68);
        assertEquals("George Bush", client2.getStudentName(stud));
        assertEquals((65.32 + 69.68 + 0) / 3, client2.getStudentTestAverage(stud));
        assertEquals(68, client2.getStudentAge(stud));
        assertEquals(school_year_t.FRESHMAN.getValue(), client2.getStudentYear(stud));
    }

    @Test
    public void test_struct_allDirection() {

        byte[] a_binary = new byte[2];
        byte[] b_binary = new byte[2];
        byte[] e_binary = new byte[2];

        for (byte i = 0; i < 2; i++) {
            a_binary[i] = i;
            b_binary[i] = (byte) (i * 2);
            e_binary[i] = i;
        }

        AllTypes a = new AllTypes(5, "Hello", enumColor.red, new C(5, 20), List.of(0, 1), List.of("Hello", "Hello"), new int[]{0, 1}, new String[]{"Hello", "Hello"}, a_binary);
        AllTypes b = new AllTypes(10, "World", enumColor.green, new C(7, 17), List.of(1, 2), List.of("World", "World"), new int[]{1, 2}, new String[]{"World", "World"}, b_binary);
        Reference<AllTypes> e = new Reference<>(new AllTypes(15, "Hello", enumColor.blue, new C(4, 14), List.of(2, 3), List.of("Hello", "Hello"), new int[]{2, 3}, new String[]{"Hello", "Hello"}, e_binary));
        client2.test_struct_allDirection(a, b, e);

        assertEquals(e.get().get_number(), a.get_number() * b.get_number());
        assertEquals(e.get().get_text(), "World");
        assertEquals(e.get().get_color(), enumColor.green);
        assertEquals(e.get().get_c().get_m(), a.get_c().get_m() * b.get_c().get_m());
        assertEquals(e.get().get_c().get_n(), a.get_c().get_n() * b.get_c().get_n());
        assertEquals(2, e.get().get_list_numbers().size());
        for (int i = 0; i < a.get_list_numbers().size(); i++) {
            assertEquals(e.get().get_list_numbers().get(i), a.get_list_numbers().get(i) * b.get_list_numbers().get(i));
            assertEquals("World", e.get().get_list_text().get(i));
        }

        for (int i = 0; i < 2; i++) {
            assertEquals(e.get().get_array_numbers()[i], a.get_array_numbers()[i] * b.get_array_numbers()[i]);
            assertEquals("World", e.get().get_list_text().get(i));
        }

        for (int i = 0; i < a.get_binary_numbers().length; i++) {
            assertEquals(e.get().get_binary_numbers()[i], a.get_binary_numbers()[i] * b.get_binary_numbers()[i]);
        }
    }

    @Test
    public void testSendingByrefMembers() {
        StructWithByrefMembers s;
        int b = 4;
        // Not sure how the test work
    }

    @AfterAll
    public void quit() {
        new CommonClient(clientManager).quit();
    }
}
